incubator/virtualcluster/experiment/pkg/scheduler/resource/virtualcluster/namespace/controller.go

```diff
@@ -0,0 +1,166 @@
+	"sigs.k8s.io/multi-tenancy/incubator/virtualcluster/experiment/pkg/scheduler/engine"
+	"sigs.k8s.io/multi-tenancy/incubator/virtualcluster/experiment/pkg/scheduler/manager"
+	"sigs.k8s.io/multi-tenancy/incubator/virtualcluster/experiment/pkg/scheduler/util"
+
+	//syncerconst "sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/syncer/constants"
+	"sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/util/listener"
+	mc "sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/util/mccontroller"
+	"sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/util/plugin"
+	"sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/util/reconciler"
+)
+
+func init() {
+	scheduler.VirtualClusterResourceRegister.Register(&plugin.Registration{
+		ID: "namespace",
+		InitFn: func(ctx *plugin.InitContext) (interface{}, error) {
+			v := ctx.Context.Value(constants.LabelSchedulerEngine)
+			if v == nil {
+				return nil, fmt.Errorf("cannot found schedulercache in context")
+			}
+			return NewNamespaceController(v.(engine.Engine), ctx.Config.(*schedulerconfig.SchedulerConfiguration))
+		},
+	})
+}
+
+type controller struct {
+	SchedulerEngine        engine.Engine
+	Config                 *schedulerconfig.SchedulerConfiguration
+	MultiClusterController *mc.MultiClusterController
+}
+
+func NewNamespaceController(schedulerEngine engine.Engine, config *schedulerconfig.SchedulerConfiguration) (manager.ResourceWatcher, error) {
+	c := &controller{
+		SchedulerEngine: schedulerEngine,
+		Config:          config,
+	}
+
+	var err error
+	c.MultiClusterController, err = mc.NewMCController(&v1.Namespace{}, &v1.NamespaceList{}, c)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (c *controller) Start(stopCh <-chan struct{}) error {
+	return c.MultiClusterController.Start(stopCh)
+}
+
+func (c *controller) GetListener() listener.ClusterChangeListener {
+	return listener.NewMCControllerListener(c.MultiClusterController)
+}
+
+func (c *controller) GetMCController() *mc.MultiClusterController {
+	return c.MultiClusterController
+}
+
+func (c *controller) Reconcile(request reconciler.Request) (reconciler.Result, error) {
+	klog.Infof("reconcile namespace %s for virtual cluster %s", request.Name, request.ClusterName)
+
+	// requeue if scheduler cache is not synchronized
+	vcName, vcNamespace, _, err := c.MultiClusterController.GetOwnerInfo(request.ClusterName)
+	if err != nil {
+		return reconciler.Result{Requeue: true}, err
+	}
+	if _, ok := scheduler.DirtyVirtualClusters.Load(fmt.Sprintf("%s/%s", vcNamespace, vcName)); ok {
+		return reconciler.Result{Requeue: true, RequeueAfter: 5 * time.Second}, fmt.Errorf("virtual cluster %s/%s is in dirty set", vcNamespace, vcName)
+	}
+
+	exists := true
+	nsObj, err := c.MultiClusterController.Get(request.ClusterName, "", request.Name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return reconciler.Result{Requeue: true}, err
+		}
+		exists = false
```

Comment:
use fast return here? or wrap the "don't exists" case in a function

lots of if condition here is such confusing